assembly/src/assembler/instruction/env_ops.rs

```diff
@@ -45,22 +42,23 @@ pub fn locaddr(
     span: &mut SpanBuilder,
     index: u16,
     context: &AssemblyContext,
-) -> Result<Option<CodeBlock>, AssemblyError> {
-    local_to_absolute_addr(span, index, context.num_proc_locals())?;
-    Ok(None)
+) -> Result<(), AssemblyError> {
+    local_to_absolute_addr(span, index, context.unwrap_current_procedure().num_locals())
 }
 
 /// Appends CALLER operation to the span which puts the hash of the function which initiated the
 /// current SYSCALL onto the stack.
 ///
 /// # Errors
 /// Returns an error if the instruction is being executed outside of kernel context.
-pub fn caller(
-    span: &mut SpanBuilder,
-    context: &AssemblyContext,
-) -> Result<Option<CodeBlock>, AssemblyError> {
-    if !context.is_kernel() {
-        return Err(AssemblyError::caller_out_of_kernel());
+pub fn caller(span: &mut SpanBuilder, context: &AssemblyContext) -> Result<(), AssemblyError> {
+    let current_procedure = context.unwrap_current_procedure();
+    if !current_procedure.is_kernel() {
+        return Err(AssemblyError::CallerOutsideOfKernel {
+            span: current_procedure.span(),
+            source_file: current_procedure.source_file(),
+        });
     }
-    span.add_op(Caller)
+    span.add_op(Caller);
+    Ok(())
```

Comment:
The error would be triggered here only if the current module is bing compiled in the `kernel` context, right? That is, if the current module executed a procedure from an imported module, and in the imported procedure contained the `caller` instruction, this would not be detected - correct?

This was also the limitation in the previous structure - just wanted to confirm if it applies to the new structure as well.

Reply:
This error basically preserves the previous one, but is actually something we detect during semantic analysis before we ever get to assembly, so it probably makes more sense for this to be an assertion now anyway.

The semantic analyzer disallows all non-kernel procedures to use `caller`. Kernel procedures (i.e. those that have "syscall" visibility), must be defined in a module with kind `ModuleKind::Kernel`, so we are able to validate things like that very early in the pipeline.

A different issue presents itself though that we discussed in that other thread yesterday, which is that since kernel modules can invoke procedures in other modules, we do need to write a pass with the ModuleGraph that validates that all of those transitively depended-upon procedures do not violate any kernel invariants. That hasn't been written yet, but is easy to implement by walking the call graph and performing the various kernel checks on each procedure. In the case of detecting invalid procedure invocations, it is even easier to verify because during semantic analysis we gather all of the procedure invocations as metadata of the procedure itself, so we don't have to revisit the body of the procedure for checks of that sort.

Reply:
I think this is fine as right now the `caller` instruction is used only in the [transaction kernel](https://github.com/0xPolygonMiden/miden-base/blob/main/miden-lib/asm/kernels/transaction/api.masm#L38) itself (and not in any of the imported modules). But yes, in a future PR we should make this more robust.